From: huawei
Author: jouyouyun <jouyouwen717@gmail.com>
Subject: huawei pangu audio patch in deepin

---
diff --git a/src/Makefile.am b/src/Makefile.am
index f4464d2..41ea0af 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -1321,11 +1321,13 @@ dist_alsaprofilesets_DATA = \
 		modules/alsa/mixer/profile-sets/kinect-audio.conf \
 		modules/alsa/mixer/profile-sets/sb-omni-surround-5.1.conf \
 		modules/alsa/mixer/profile-sets/steelseries-arctis-usb-audio.conf \
-		modules/alsa/mixer/profile-sets/dell-dock-tb16-usb-audio.conf
+		modules/alsa/mixer/profile-sets/dell-dock-tb16-usb-audio.conf \
+		modules/alsa/mixer/profile-sets/pulse-huawei.conf
 
 if HAVE_UDEV
 dist_udevrules_DATA = \
-		modules/alsa/mixer/profile-sets/90-pulseaudio.rules
+		modules/alsa/mixer/profile-sets/90-pulseaudio.rules \
+		modules/alsa/mixer/profile-sets/91-pulseaudio-huawei.rules
 endif
 
 dist_alsapaths_DATA = \
@@ -1366,7 +1368,15 @@ dist_alsapaths_DATA = \
 		modules/alsa/mixer/paths/hdmi-output-7.conf \
 		modules/alsa/mixer/paths/steelseries-arctis-input.conf \
 		modules/alsa/mixer/paths/steelseries-arctis-output-mono.conf \
-		modules/alsa/mixer/paths/steelseries-arctis-output-stereo.conf
+		modules/alsa/mixer/paths/steelseries-arctis-output-stereo.conf \
+		modules/alsa/mixer/paths/analog-input-headset-mic-huawei.conf \
+		modules/alsa/mixer/paths/analog-input-internal-mic-huawei.conf \
+		modules/alsa/mixer/paths/analog-input-linein-huawei.conf \
+		modules/alsa/mixer/paths/analog-input-rear-mic-huawei.conf \
+		modules/alsa/mixer/paths/analog-output-headphones-huawei.conf \
+		modules/alsa/mixer/paths/analog-output-headphones-huawei-2.conf \
+		modules/alsa/mixer/paths/analog-output-speaker-huawei.conf \
+		modules/alsa/mixer/paths/analog-output-speaker-huawei-2.conf
 
 endif
 
diff --git a/src/modules/alsa/mixer/paths/analog-input-headset-mic-huawei.conf b/src/modules/alsa/mixer/paths/analog-input-headset-mic-huawei.conf
new file mode 100644
index 0000000..e466b49
--- /dev/null
+++ b/src/modules/alsa/mixer/paths/analog-input-headset-mic-huawei.conf
@@ -0,0 +1,44 @@
+# This file is part of PulseAudio.
+#
+# PulseAudio is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as
+# published by the Free Software Foundation; either version 2.1 of the
+# License, or (at your option) any later version.
+#
+# PulseAudio is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with PulseAudio; if not, see <http://www.gnu.org/licenses/>.
+
+; For devices where a 'Line' element exists
+;
+; See analog-output.conf.common for an explanation on the directives
+
+[General]
+priority = 90
+description-key = analog-input-microphone-headset
+
+[Jack Headset - Input]
+required-any = any
+
+[Element Rear Mic]
+switch = off
+volume = off
+
+[Element Line]
+switch = off
+volume = off
+
+[Element Mic]
+switch = off
+volume = off
+
+[Element Headset Mic]
+required-any = any
+switch = on
+volume = merge
+override-map.1 = all
+override-map.2 = all-left,all-right
diff --git a/src/modules/alsa/mixer/paths/analog-input-internal-mic-huawei.conf b/src/modules/alsa/mixer/paths/analog-input-internal-mic-huawei.conf
new file mode 100644
index 0000000..3762713
--- /dev/null
+++ b/src/modules/alsa/mixer/paths/analog-input-internal-mic-huawei.conf
@@ -0,0 +1,42 @@
+# This file is part of PulseAudio.
+#
+# PulseAudio is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as
+# published by the Free Software Foundation; either version 2.1 of the
+# License, or (at your option) any later version.
+#
+# PulseAudio is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with PulseAudio; if not, see <http://www.gnu.org/licenses/>.
+
+; For devices where a 'Internal Mic' or 'Internal Mic Boost' element exists
+; 'Int Mic' and 'Int Mic Boost' are for compatibility with kernels < 2.6.38
+;
+; See analog-output.conf.common for an explanation on the directives
+
+[General]
+priority = 87
+description-key = analog-input-microphone-internal
+
+[Jack Headset Mic]
+state.plugged = no
+state.unplugged = unknown
+
+[Jack Mic]
+required-any = any
+
+[Element Headset Mic]
+switch = off
+volume = off
+
+[Element Mic]
+required-any = any
+switch = on
+volume = merge
+override-map.1 = all
+override-map.2 = all-left,all-right
+
diff --git a/src/modules/alsa/mixer/paths/analog-input-linein-huawei.conf b/src/modules/alsa/mixer/paths/analog-input-linein-huawei.conf
new file mode 100644
index 0000000..0f1e1a1
--- /dev/null
+++ b/src/modules/alsa/mixer/paths/analog-input-linein-huawei.conf
@@ -0,0 +1,50 @@
+# This file is part of PulseAudio.
+#
+# PulseAudio is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as
+# published by the Free Software Foundation; either version 2.1 of the
+# License, or (at your option) any later version.
+#
+# PulseAudio is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with PulseAudio; if not, see <http://www.gnu.org/licenses/>.
+
+; For devices where a 'Line' element exists
+;
+; See analog-output.conf.common for an explanation on the directives
+
+[General]
+priority = 88
+description-key = analog-input-linein
+
+[Jack Line - Input]
+required-any = any
+
+[Jack Headset - Input]
+state.plugged = no
+state.unplugged = unknown
+
+[Jack Mic - Input]
+state.plugged = no
+state.unplugged = unknown
+
+[Element Headset Mic]
+switch = off
+volume = off
+
+[Element Rear Mic]
+switch = off
+volume = off
+
+[Element Line]
+required-any = any
+switch = on
+volume = merge
+override-map.1 = all
+override-map.2 = all-left,all-right
+
+
diff --git a/src/modules/alsa/mixer/paths/analog-input-rear-mic-huawei.conf b/src/modules/alsa/mixer/paths/analog-input-rear-mic-huawei.conf
new file mode 100644
index 0000000..c5d7646
--- /dev/null
+++ b/src/modules/alsa/mixer/paths/analog-input-rear-mic-huawei.conf
@@ -0,0 +1,45 @@
+# This file is part of PulseAudio.
+#
+# PulseAudio is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as
+# published by the Free Software Foundation; either version 2.1 of the
+# License, or (at your option) any later version.
+#
+# PulseAudio is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with PulseAudio; if not, see <http://www.gnu.org/licenses/>.
+
+; For devices where a 'Line' element exists
+;
+; See analog-output.conf.common for an explanation on the directives
+
+[General]
+priority = 89
+description-key = analog-input-microphone-rear
+
+[Jack Mic - Input]
+required-any = any
+
+[Jack Headset - Input]
+state.plugged = no
+state.unplugged = unknown
+
+[Element Headset Mic]
+switch = off
+volume = off
+
+[Element Line]
+switch = off
+volume = off
+
+[Element Rear Mic]
+required-any = any
+switch = on
+volume = merge
+override-map.1 = all
+override-map.2 = all-left,all-right
+
diff --git a/src/modules/alsa/mixer/paths/analog-output-headphones-huawei-2.conf b/src/modules/alsa/mixer/paths/analog-output-headphones-huawei-2.conf
new file mode 100644
index 0000000..b5d9e21
--- /dev/null
+++ b/src/modules/alsa/mixer/paths/analog-output-headphones-huawei-2.conf
@@ -0,0 +1,47 @@
+# This file is part of PulseAudio.
+#
+# PulseAudio is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as
+# published by the Free Software Foundation; either version 2.1 of the
+# License, or (at your option) any later version.
+#
+# PulseAudio is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with PulseAudio; if not, see <http://www.gnu.org/licenses/>.
+
+; Path for mixers that have a 'Headphone' control
+;
+; See analog-output.conf.common for an explanation on the directives
+
+[General]
+priority = 89
+description-key = analog-output-headphones
+
+[Properties]
+device.icon_name = audio-headphones
+
+[Jack Headphone - Output]
+required-any = any
+
+[Jack Headset - Output]
+state.plugged = no
+state.unplugged = unknown
+
+[Element Headset]
+switch = off
+volume = off
+
+[Element Desktop Speaker]
+switch = off
+volume = off
+
+[Element Headphone2]
+required = any
+switch = on
+volume = merge
+override-map.1 = all
+override-map.2 = all-left,all-right
diff --git a/src/modules/alsa/mixer/paths/analog-output-headphones-huawei.conf b/src/modules/alsa/mixer/paths/analog-output-headphones-huawei.conf
new file mode 100644
index 0000000..42384de
--- /dev/null
+++ b/src/modules/alsa/mixer/paths/analog-output-headphones-huawei.conf
@@ -0,0 +1,50 @@
+# This file is part of PulseAudio.
+#
+# PulseAudio is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as
+# published by the Free Software Foundation; either version 2.1 of the
+# License, or (at your option) any later version.
+#
+# PulseAudio is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with PulseAudio; if not, see <http://www.gnu.org/licenses/>.
+
+; Path for mixers that have a 'Headphone' control
+;
+; See analog-output.conf.common for an explanation on the directives
+
+[General]
+priority = 91
+description-key = analog-output-headphones
+
+[Properties]
+device.icon_name = audio-headphones
+
+[Jack Headset - Output]
+required-any = any
+
+[Jack Headphone]
+required-any = any
+
+[Element Headphone2]
+switch = off
+volume = off
+
+[Element Desktop Speaker]
+switch = off
+volume = off
+
+[Element Speaker]
+switch = off
+volume = off
+
+[Element Headset]
+required = any
+switch = on
+volume = merge
+override-map.1 = all
+override-map.2 = all-left,all-right
diff --git a/src/modules/alsa/mixer/paths/analog-output-speaker-huawei-2.conf b/src/modules/alsa/mixer/paths/analog-output-speaker-huawei-2.conf
new file mode 100644
index 0000000..79013ba
--- /dev/null
+++ b/src/modules/alsa/mixer/paths/analog-output-speaker-huawei-2.conf
@@ -0,0 +1,51 @@
+# This file is part of PulseAudio.
+#
+# PulseAudio is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as
+# published by the Free Software Foundation; either version 2.1 of the
+# License, or (at your option) any later version.
+#
+# PulseAudio is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with PulseAudio; if not, see <http://www.gnu.org/licenses/>.
+
+; Path for mixers that have a 'Speaker' control
+;
+; See analog-output.conf.common for an explanation on the directives
+
+[General]
+priority = 100
+description-key = analog-output-speaker
+
+[Properties]
+device.icon_name = audio-speakers
+
+[Jack Headset - Output]
+state.plugged = no
+state.unplugged = unknown
+
+[Jack Headphone - Output]
+state.plugged = no
+state.unplugged = unknown
+
+[Jack Desktop Speaker - Output]
+required-any = any
+
+[Element Headset]
+switch = off
+volume = off
+
+[Element Headphone2]
+switch = off
+volume = off
+
+[Element Desktop Speaker]
+required-any = any
+switch = on
+volume = merge
+override-map.1 = all
+override-map.2 = all-left,all-right
diff --git a/src/modules/alsa/mixer/paths/analog-output-speaker-huawei.conf b/src/modules/alsa/mixer/paths/analog-output-speaker-huawei.conf
new file mode 100644
index 0000000..eb89641
--- /dev/null
+++ b/src/modules/alsa/mixer/paths/analog-output-speaker-huawei.conf
@@ -0,0 +1,49 @@
+# This file is part of PulseAudio.
+#
+# PulseAudio is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as
+# published by the Free Software Foundation; either version 2.1 of the
+# License, or (at your option) any later version.
+#
+# PulseAudio is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with PulseAudio; if not, see <http://www.gnu.org/licenses/>.
+
+; Path for mixers that have a 'Speaker' control
+;
+; See analog-output.conf.common for an explanation on the directives
+
+[General]
+priority = 87
+description-key = analog-output-speaker
+
+[Properties]
+device.icon_name = audio-speakers
+
+[Jack Headphone]
+state.plugged = no
+state.unplugged = unknown
+
+[Jack Speaker]
+required-any = any
+state.plugged = unknown
+state.unplugged = unknown
+
+
+[Element Headphone]
+switch = off
+volume = off
+
+[Element Speaker]
+required-any = any
+switch = on
+volume = merge
+override-map.1 = all
+override-map.2 = all-left,all-right
+
+
+
diff --git a/src/modules/alsa/mixer/profile-sets/91-pulseaudio-huawei.rules b/src/modules/alsa/mixer/profile-sets/91-pulseaudio-huawei.rules
new file mode 100644
index 0000000..622f1fd
--- /dev/null
+++ b/src/modules/alsa/mixer/profile-sets/91-pulseaudio-huawei.rules
@@ -0,0 +1,28 @@
+# do not edit this file, it will be overwritten on update
+
+# This file is part of PulseAudio.
+#
+# PulseAudio is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as
+# published by the Free Software Foundation; either version 2.1 of the
+# License, or (at your option) any later version.
+#
+# PulseAudio is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with PulseAudio; if not, see <http://www.gnu.org/licenses/>.
+
+SUBSYSTEM!="sound", GOTO="pulseaudio_end"
+ACTION!="change", GOTO="pulseaudio_end"
+KERNEL!="card*", GOTO="pulseaudio_end"
+SUBSYSTEMS=="usb", GOTO="pulseaudio_check_usb"
+
+GOTO="pulseaudio_end"
+
+LABEL="pulseaudio_check_usb"
+ATTRS{idVendor}=="0bda", ATTRS{idProduct}=="49ab", ENV{PULSE_PROFILE_SET}="pulse-huawei.conf"
+
+LABEL="pulseaudio_end"
diff --git a/src/modules/alsa/mixer/profile-sets/pulse-huawei.conf b/src/modules/alsa/mixer/profile-sets/pulse-huawei.conf
new file mode 100644
index 0000000..4e86a67
--- /dev/null
+++ b/src/modules/alsa/mixer/profile-sets/pulse-huawei.conf
@@ -0,0 +1,132 @@
+# This file is part of PulseAudio.
+#
+# PulseAudio is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as
+# published by the Free Software Foundation; either version 2.1 of the
+# License, or (at your option) any later version.
+#
+# PulseAudio is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with PulseAudio; if not, see <http://www.gnu.org/licenses/>.
+
+; Default profile definitions for the ALSA backend of PulseAudio. This
+; is used as fallback for all cards that have no special mapping
+; assigned (and should be good enough for the vast majority of
+; cards). If you want to assign a different profile set than this one
+; to a device, either set the udev property PULSE_PROFILE_SET for the
+; card, or use the "profile_set" module argument when loading
+; module-alsa-card.
+;
+; So what is this about? Simply, what we do here is map ALSA devices
+; to how they are exposed in PA. We say which ALSA device string to
+; use to open a device, which channel mapping to use then, and which
+; mixer path to use. This is encoded in a 'mapping'. Multiple of these
+; mappings can be bound together in a 'profile' which is then directly
+; exposed in the UI as a card profile. Each mapping assigned to a
+; profile will result in one sink/source to be created if the profile
+; is selected for the card.
+;
+; Additionally, the path set configuration files can describe the
+; decibel values assigned to the steps of the volume elements. This
+; can be used to work around situations when the alsa driver doesn't
+; provide any decibel information, or when the information is
+; incorrect.
+
+
+; [General]
+; auto-profiles = no | yes                  # Instead of defining all profiles manually, autogenerate
+;                                           # them by combining every input mapping with every output mapping.
+;
+; [Mapping id]
+; device-strings = ...                      # ALSA device string. %f will be replaced by the card identifier.
+; channel-map = ...                         # Channel mapping to use for this device
+; description = ...
+; paths-input = ...                         # A list of mixer paths to use. Every path in this list will be probed.
+;                                           # If multiple are found to be working they will be available as device ports
+; paths-output = ...
+; element-input = ...                       # Instead of configuring a full mixer path simply configure a single
+;                                           # mixer element for volume/mute handling
+; element-output = ...
+; priority = ...
+; direction = any | input | output          # Only useful for?
+;
+; exact-channels = yes | no                 # If no, and the exact number of channels is not supported,
+;                                           # allow device to be opened with another channel count
+; fallback = no | yes                       # This mapping will only be considered if all non-fallback mappings fail
+; [Profile id]
+; input-mappings = ...                      # Lists mappings for sources on this profile, those mapping must be
+;                                           # defined in this file too
+; output-mappings = ...                     # Lists mappings for sinks on this profile, those mappings must be
+;                                           # defined in this file too
+; description = ...
+; priority = ...                            # Numeric value to deduce priority for this profile
+; skip-probe = no | yes                     # Skip probing for availability? If this is yes then this profile
+;                                           # will be assumed as working without probing. Makes initialization
+;                                           # a bit faster but only works if the card is really known well.
+;
+; fallback = no | yes                       # This profile will only be considered if all non-fallback profiles fail
+; [DecibelFix element]                      # Decibel fixes can be used to work around missing or incorrect dB
+;                                           # information from alsa. A decibel fix is a table that maps volume steps
+;                                           # to decibel values for one volume element. The "element" part in the
+;                                           # section title is the name of the volume element.
+;                                           #
+;                                           # NOTE: This feature is meant just as a help for figuring out the correct
+;                                           # decibel values. PulseAudio is not the correct place to maintain the
+;                                           # decibel mappings!
+;                                           #
+;                                           # If you need this feature, then you should make sure that when you have
+;                                           # the correct values figured out, the alsa driver developers get informed
+;                                           # too, so that they can fix the driver.
+;
+; db-values = ...                           # The option value consists of pairs of step numbers and decibel values.
+;                                           # The pairs are separated with whitespace, and steps are separated from
+;                                           # the corresponding decibel values with a colon. The values must be in an
+;                                           # increasing order. Here's an example of a valid string:
+;                                           #
+;                                           #     "0:-40.50  1:-38.70  3:-33.00  11:0"
+;                                           #
+;                                           # The lowest step imposes a lower limit for hardware volume and the
+;                                           # highest step correspondingly imposes a higher limit. That means that
+;                                           # that the mixer will never be set outside those values - the rest of the
+;                                           # volume scale is done using software volume.
+;                                           #
+;                                           # As can be seen in the example, you don't need to specify a dB value for
+;                                           # each step. The dB values for skipped steps will be linearly interpolated
+;                                           # using the nearest steps that are given.
+
+[General]
+auto-profiles = yes
+
+[Mapping analog-stereo]
+device-strings = front:%f
+channel-map = left,right
+paths-output = analog-output-headphones-huawei analog-output-headphones-huawei-2 analog-output-speaker-huawei
+paths-input = analog-input-rear-mic-huawei analog-input-headset-mic-huawei analog-input-internal-mic-huawei 
+priority = 10
+
+[Mapping analog-stereo-input]
+device-strings = hw:%f,1
+channel-map = left,right
+paths-input = analog-input-linein-huawei
+direction = input
+priority = 9
+
+[Mapping analog-stereo-output]
+device-strings = hw:%f,1
+channel-map = left,right
+paths-output = analog-output-speaker-huawei-2
+direction = output
+priority = 11
+
+[DecibelFix Headset Mic]
+db-values = 0:-65.62 47:0
+
+[DecibelFix Rear Mic]
+db-values = 0:-65.62 47:0
+
+[DecibelFix Line]
+db-values = 0:-65.62 47:0
