diff --git a/src/modules/alsa/mixer/paths/st2557-pc-usb-audio.conf b/src/modules/alsa/mixer/paths/st2557-pc-usb-audio.conf
new file mode 100644
index 0000000..f66a9e4
--- /dev/null
+++ b/src/modules/alsa/mixer/paths/st2557-pc-usb-audio.conf
@@ -0,0 +1,51 @@
+# This file is part of PulseAudio.
+#
+# PulseAudio is free software; you can redistribute it and/or modify
+# it under the terms of the GNU Lesser General Public License as
+# published by the Free Software Foundation; either version 2.1 of the
+# License, or (at your option) any later version.
+#
+# PulseAudio is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with PulseAudio; if not, see <http://www.gnu.org/licenses/>.
+
+[General]
+auto-profiles = yes
+
+[Mapping analog-stereo-headphone]
+description = Headphone
+paths-output = analog-output analog-output-headphones
+device-strings = hw:%f,0
+channel-map = left,right
+direction = output
+
+[Mapping analog-stereo-linein]
+description = Linein
+paths-input = analog-input analog-input-linein
+device-strings = hw:%f,2
+channel-map = left,right
+direction = input
+
+[Mapping analog-stereo-rear-mic]
+description = Rear Microphone
+paths-input = analog-input analog-input-rear-mic
+device-strings = hw:%f,1
+channel-map = left,right
+direction = input
+
+[Mapping analog-stereo-front-mic]
+description = Front Microphone
+paths-input = analog-input analog-input-front-mic
+device-strings = hw:%f,0
+channel-map = left
+direction = input
+
+[Profile output:analog-stereo-headphone+input:analog-stereo-all]
+description = Huaqin ST2557-PC
+output-mappings = analog-stereo-headphone
+input-mappings = analog-stereo-linein analog-stereo-rear-mic analog-stereo-front-mic
+skip-probe = yes
diff --git a/src/modules/alsa/mixer/profile-sets/90-pulseaudio.rules b/src/modules/alsa/mixer/profile-sets/90-pulseaudio.rules
index 264bfdb..de75d74 100644
--- a/src/modules/alsa/mixer/profile-sets/90-pulseaudio.rules
+++ b/src/modules/alsa/mixer/profile-sets/90-pulseaudio.rules
@@ -107,4 +107,6 @@ ATTRS{idVendor}=="041e", ATTRS{idProduct}=="322c", ENV{PULSE_PROFILE_SET}="sb-om
 ATTRS{idVendor}=="0bda", ATTRS{idProduct}=="4014", ENV{PULSE_PROFILE_SET}="dell-dock-tb16-usb-audio.conf"
 ATTRS{idVendor}=="1038", ATTRS{idProduct}=="1260", ENV{PULSE_PROFILE_SET}="steelseries-arctis-usb-audio.conf"
 
+ATTRS{idVendor}=="0bda", ATTRS{idProduct}=="4987", ENV{PULSE_PROFILE_SET}="st2557-pc-usb-audio.conf"
+
 LABEL="pulseaudio_end"
